app/assets/javascripts/discourse/app/lib/model-transformers.js

```diff
@@ -0,0 +1,34 @@
+import { consolePrefix } from "discourse/lib/source-identifier";
+
+let modelTransformersMap = {};
+
+export function registerModelTransformer(modelName, func) {
+  if (!modelTransformersMap[modelName]) {
+    modelTransformersMap[modelName] = [];
+  }
+  const transformer = {
+    prefix: consolePrefix(),
+    execute: func,
+  };
+  modelTransformersMap[modelName].push(transformer);
+}
+
+export async function applyModelTransformations(modelName, models) {
+  for (const transformer of modelTransformersMap[modelName] || []) {
+    try {
+      await transformer.execute(models);
+    } catch (err) {
+      // eslint-disable-next-line no-console
+      console.error(
+        transformer.prefix,
+        `transformer for the \`${modelName}\` model failed with:`,
+        err,
+        err.stack
+      );
+    }
+  }
+}
+
+export function resetModelTransformers() {
+  modelTransformersMap = {};
```

Comment:
@CvX For objects, do we need to avoid reassign it? Was wondering if this will result in the same problem as you encountered in https://github.com/discourse/discourse/commit/349f6e18f380116eac4f08457c864a4eb8428292